File: copad-core/src/session.rs
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

// ============================================================================
// v3 session model (decision #64 — SSH-first / Model 1 simplification).
//
// A flat tab list; each tab is a split tree of TYPED panes. Persistence is
// LAYOUT-ONLY: tab list + split layout (orientation + normalized ratio) + each
// terminal's cwd. Terminals restore as FRESH shells — process/scrollback
// persistence is the user's own tmux, not copad's (decision #63). This replaces
// the v1 (terminal-only) and v2 (workspace → sub-tab → pane, tmux-backed) models,
// both deleted. Old v1/v2 files are rejected by the loader and a fresh session
// starts — the accepted wipe-fresh transition.
// ============================================================================

pub const SESSION_VERSION: u32 = 3;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Session {
    pub version: u32,
    pub tabs: Vec<TabSnap>,
    pub current_tab: usize,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct TabSnap {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub custom_title: Option<String>,
    pub root: SplitSnap,
}

/// A split tree: a typed leaf pane, or a branch of two subtrees. `ratio` is a
/// normalized 0..1 divider position (never platform pixels) so a split's size
/// restores faithfully across Linux and macOS.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SplitSnap {
    Leaf {
        content: PaneContent,
    },
    Branch {
        orientation: SplitOrientation,
        ratio: f32,
        first: Box<SplitSnap>,
        second: Box<SplitSnap>,
    },
}

/// Typed pane content. Non-terminal panes are copad-native (tmux never backs
/// them). A terminal persists only its cwd — it restores as a fresh shell.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PaneContent {
    Terminal {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        cwd: Option<String>,
    },
    Webview {
        /// The pane's URL under the active `[browser] restore` policy. Under the
        /// default `origin` policy this is the canonical origin only — never the
        /// live URL (creds / OAuth codes / signed URLs). Under the opt-in `url` /
        /// `full` policies it carries more; see `browser::restore`.
        url: String,
        /// Full browser state for this pane: tab list, active index, profile.
        /// Additive and `#[serde(default)]`, so v3 files written before the
        /// Browser Workbench keep loading and `SESSION_VERSION` does not move.
        ///
        /// **Downgrade is lossy but never corrupt** (codex plan r2-I4): an older
        /// binary round-trips the file and drops this field, degrading the pane
        /// to today's single-URL restore. That is the accepted behaviour — the
        /// alternative (a version bump) would wipe every user's whole layout.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        pane: Option<crate::browser::BrowserPaneSnap>,
    },
    Plugin {
        name: String,
        /// The specific panel within a multi-panel plugin, so restore targets
        /// the original panel rather than the plugin's default.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        panel_name: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        version: Option<String>,
    },
    /// The agent-status cockpit pane (Linux). No persisted state beyond its kind.
    Cockpit,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SplitOrientation {
    Horizontal,
    Vertical,
}

/// Divider-ratio clamp band so a persisted extreme / corrupt value can't
/// collapse a pane below usability.
pub const MIN_RATIO: f32 = 0.05;
pub const MAX_RATIO: f32 = 0.95;

pub fn clamp_ratio(r: f32) -> f32 {
    if r.is_finite() {
        r.clamp(MIN_RATIO, MAX_RATIO)
    } else {
        0.5
    }
}

/// Clamp every divider ratio in a tree (defense against hand-edited / corrupt
/// files) so downstream layout never sees an out-of-band value.
pub fn normalize_ratios(snap: &mut SplitSnap) {
    if let SplitSnap::Branch {
        ratio,
        first,
        second,
        ..
    } = snap
    {
        *ratio = clamp_ratio(*ratio);
        normalize_ratios(first);
        normalize_ratios(second);
    }
}

pub fn session_path() -> PathBuf {
    crate::paths::state_dir().join("session.json")
}

pub fn load() -> Option<Session> {
    let raw = std::fs::read_to_string(session_path()).ok()?;
    parse_session(&raw)
}

/// Parse + validate a session document. Split from `load` so it is unit-testable
/// without the filesystem. Rejects unknown versions outright (a best-effort parse
/// of a foreign schema risks a half-restored state worse than starting fresh) —
/// this is what makes old v1/v2 files fall back to a fresh session (wipe-fresh).
pub fn parse_session(raw: &str) -> Option<Session> {
    let mut session: Session = serde_json::from_str(raw)
        .map_err(|e| eprintln!("[copad] session parse failed: {e}"))
        .ok()?;
    if session.version != SESSION_VERSION {
        eprintln!(
            "[copad] session version mismatch (file={}, expected={SESSION_VERSION}) — ignoring",
            session.version
        );
        return None;
    }
    if session.tabs.is_empty() {
        return None;
    }
    for t in &mut session.tabs {
        normalize_ratios(&mut t.root);
    }
    Some(session)
}

/// Remove any persisted session file. Called when the closing window has no
/// panels left so a stale snapshot doesn't restore on next launch.
pub fn clear() {
    let path = session_path();
    if let Err(e) = std::fs::remove_file(&path)
        && e.kind() != std::io::ErrorKind::NotFound
    {
        eprintln!("[copad] session clear failed: {e}");
    }
}

pub fn save(session: &Session) {
    let path = session_path();
    let Some(parent) = path.parent() else { return };
    if let Err(e) = std::fs::create_dir_all(parent) {
        eprintln!(
            "[copad] session save: mkdir {} failed: {e}",
            parent.display()
        );
        return;
    }
    let json = match serde_json::to_string_pretty(session) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("[copad] session serialize failed: {e}");
            return;
        }
    };
    let tmp = path.with_extension("json.tmp");
    if let Err(e) = std::fs::write(&tmp, json) {
        eprintln!("[copad] session write {} failed: {e}", tmp.display());
        return;
    }
    if let Err(e) = std::fs::rename(&tmp, &path) {
        eprintln!("[copad] session rename failed: {e}");
        let _ = std::fs::remove_file(&tmp);
    }
}

/// Walk a `SplitSnap` and return the cwd of the leftmost (DFS pre-order)
/// TERMINAL leaf, skipping non-terminal (webview / plugin / cockpit) leaves so a
/// non-terminal-first tree still seeds a later terminal's cwd. Used at restore:
/// the cwd of the first terminal seeds the panel that opens the tab.
pub fn leftmost_cwd(snap: &SplitSnap) -> Option<String> {
    match snap {
        SplitSnap::Leaf {
            content: PaneContent::Terminal { cwd },
        } => cwd.clone(),
        SplitSnap::Leaf { .. } => None,
        SplitSnap::Branch { first, second, .. } => {
            leftmost_cwd(first).or_else(|| leftmost_cwd(second))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn term(cwd: &str) -> SplitSnap {
        SplitSnap::Leaf {
            content: PaneContent::Terminal {
                cwd: Some(cwd.to_string()),
            },
        }
    }

    fn webview(url: &str) -> SplitSnap {
        SplitSnap::Leaf {
            content: PaneContent::Webview {
                url: url.to_string(),
                pane: None,
            },
        }
    }

    fn branch(o: SplitOrientation, ratio: f32, first: SplitSnap, second: SplitSnap) -> SplitSnap {
        SplitSnap::Branch {
            orientation: o,
            ratio,
            first: Box::new(first),
            second: Box::new(second),
        }
    }

    #[test]
    fn round_trips_mixed_typed_tree() {
        let s = Session {
            version: SESSION_VERSION,
            current_tab: 1,
            tabs: vec![
                TabSnap {
                    custom_title: Some("editor".into()),
                    root: branch(
                        SplitOrientation::Horizontal,
                        0.6,
                        branch(
                            SplitOrientation::Vertical,
                            0.5,
                            term("/a"),
                            webview("https://example.com"),
                        ),
                        SplitSnap::Leaf {
                            content: PaneContent::Plugin {
                                name: "kb".into(),
                                panel_name: Some("notes".into()),
                                version: Some("1".into()),
                            },
                        },
                    ),
                },
                TabSnap {
                    custom_title: None,
                    root: SplitSnap::Leaf {
                        content: PaneContent::Cockpit,
                    },
                },
            ],
        };
        let json = serde_json::to_string(&s).unwrap();
        let back: Session = serde_json::from_str(&json).unwrap();
        assert_eq!(s, back);
    }

    #[test]
    fn parse_rejects_old_and_unknown_versions() {
        // v1 doc (version 1) and v2 doc (version 2) both fall back to fresh.
        let v1 = r#"{"version":1,"tabs":[{"custom_title":null,"root":{"type":"terminal","cwd":"/x"}}],"current_tab":0}"#;
        assert!(parse_session(v1).is_none(), "old v1 file rejected → fresh");
        let v2 = r#"{"version":2,"sessions":[]}"#;
        assert!(parse_session(v2).is_none(), "old v2 file rejected → fresh");
        let future = r#"{"version":999,"tabs":[],"current_tab":0}"#;
        assert!(
            parse_session(future).is_none(),
            "unknown future version rejected"
        );
    }

    #[test]
    fn parse_rejects_empty_tab_list() {
        let doc = r#"{"version":3,"tabs":[],"current_tab":0}"#;
        assert!(parse_session(doc).is_none());
    }

    #[test]
    fn parse_clamps_out_of_band_ratio() {
        let doc = r#"{"version":3,"current_tab":0,"tabs":[{"root":{"type":"branch","orientation":"horizontal","ratio":9.0,"first":{"type":"leaf","content":{"kind":"terminal"}},"second":{"type":"leaf","content":{"kind":"terminal"}}}}]}"#;
        let s = parse_session(doc).expect("v3 doc parses");
        match &s.tabs[0].root {
            SplitSnap::Branch { ratio, .. } => {
                assert!(
                    *ratio <= MAX_RATIO && *ratio >= MIN_RATIO,
                    "ratio clamped on load"
                );
            }
            _ => panic!("expected branch"),
        }
    }

    #[test]
    fn leftmost_cwd_skips_non_terminal_leaves() {
        // A webview-first branch must still find the terminal's cwd in `second`.
        let s = branch(
            SplitOrientation::Horizontal,
            0.5,
            webview("https://x"),
            term("/deep"),
        );
        assert_eq!(leftmost_cwd(&s), Some("/deep".to_string()));
        // Nested: leftmost terminal wins.
        let s = branch(
            SplitOrientation::Horizontal,
            0.5,
            branch(
                SplitOrientation::Vertical,
                0.5,
                webview("https://y"),
                term("/a"),
            ),
            term("/b"),
        );
        assert_eq!(leftmost_cwd(&s), Some("/a".to_string()));
    }

    #[test]
    fn leftmost_cwd_none_when_no_terminal() {
        let s = branch(
            SplitOrientation::Horizontal,
            0.5,
            webview("https://x"),
            SplitSnap::Leaf {
                content: PaneContent::Cockpit,
            },
        );
        assert_eq!(leftmost_cwd(&s), None);
    }

    #[test]
    fn round_trip_through_parse_after_save_shape() {
        let s = Session {
            version: SESSION_VERSION,
            current_tab: 0,
            tabs: vec![TabSnap {
                custom_title: None,
                root: term("/home/x"),
            }],
        };
        let json = serde_json::to_string_pretty(&s).unwrap();
        let back = parse_session(&json).expect("our own output parses back");
        assert_eq!(s, back);
    }

    // ---- Browser Workbench: the additive `pane` field (plan B1) ----

    #[test]
    fn a_pre_workbench_v3_webview_pane_still_loads() {
        // The whole reason `pane` is `#[serde(default)]` rather than a version
        // bump: every existing user's layout file lacks the field.
        let raw = r#"{
            "version": 3,
            "current_tab": 0,
            "tabs": [{ "root": {
                "type": "leaf",
                "content": { "kind": "webview", "url": "https://github.com" }
            }}]
        }"#;
        let s = parse_session(raw).expect("legacy v3 webview pane must still load");
        let SplitSnap::Leaf { content } = &s.tabs[0].root else {
            panic!("expected a leaf");
        };
        assert_eq!(
            content,
            &PaneContent::Webview {
                url: "https://github.com".into(),
                pane: None,
            }
        );
    }

    #[test]
    fn a_pane_without_browser_state_serializes_exactly_as_before() {
        // `skip_serializing_if` keeps the on-disk shape byte-identical for
        // users who never open a second browser tab, so an older binary reading
        // the file sees nothing new at all.
        let json = serde_json::to_string(&PaneContent::Webview {
            url: "https://e.com".into(),
            pane: None,
        })
        .unwrap();
        assert!(!json.contains("pane"), "{json}");
    }

    #[test]
    fn browser_tabs_round_trip_through_a_session_document() {
        use crate::browser::{BrowserPaneSnap, BrowserTabSnap};

        let pane = BrowserPaneSnap {
            tabs: vec![
                BrowserTabSnap::new("tab-a", "https://github.com/o/r/pull/42"),
                BrowserTabSnap::new("tab-b", "https://example.com"),
            ],
            active: 1,
            profile: "default".into(),
        };
        let session = Session {
            version: SESSION_VERSION,
            current_tab: 0,
            tabs: vec![TabSnap {
                custom_title: None,
                root: SplitSnap::Leaf {
                    content: PaneContent::Webview {
                        url: "https://github.com/o/r/pull/42".into(),
                        pane: Some(pane),
                    },
                },
            }],
        };
        let json = serde_json::to_string(&session).unwrap();
        assert_eq!(parse_session(&json).expect("round-trips"), session);
    }

    #[test]
    fn an_older_binary_downgrades_a_browser_pane_lossily_but_not_corruptly() {
        // codex plan r2-I4. Simulate the old binary by parsing into a shape that
        // has no `pane` field and re-serializing: the layout survives, the tab
        // list is dropped, and nothing about the document becomes invalid.
        use crate::browser::{BrowserPaneSnap, BrowserTabSnap};

        let with_tabs = PaneContent::Webview {
            url: "https://github.com".into(),
            pane: Some(BrowserPaneSnap {
                tabs: vec![BrowserTabSnap::new("tab-a", "https://github.com")],
                active: 0,
                profile: "default".into(),
            }),
        };
        let json = serde_json::to_string(&with_tabs).unwrap();

        // An old binary's view of the same variant.
        #[derive(serde::Serialize, serde::Deserialize)]
        #[serde(tag = "kind", rename_all = "snake_case")]
        enum OldPaneContent {
            Webview { url: String },
        }
        let old: OldPaneContent = serde_json::from_str(&json).expect("old binary still parses");
        let rewritten = serde_json::to_string(&old).unwrap();

        // Re-read with the current model: valid, and degraded to a single URL.
        let back: PaneContent = serde_json::from_str(&rewritten).unwrap();
        assert_eq!(
            back,
            PaneContent::Webview {
                url: "https://github.com".into(),
                pane: None,
            }
        );
    }
}

File: copad-macos/Sources/Copad/Session.swift
import CCopadFFI
import CopadCore
import Foundation

/// Persisted tab/split layout (v3 — decision #64). Wire shape mirrors
/// `copad_core::session::Session` exactly: a flat tab list, each tab a split
/// tree of TYPED panes (`type`-tagged `SplitSnap`, `kind`-tagged `PaneContent`,
/// lowercase orientation, normalized `ratio`). Persistence is LAYOUT-ONLY:
/// terminals persist only their cwd and restore as fresh shells — process
/// persistence is the user's own tmux. Load / save / clear are owned by core and
/// reached through `copad-ffi`; this file only encodes / decodes JSON and walks
/// the in-memory tree (`leftmostCwd`).
enum Session {
    static let version: Int = 3

    /// Returns nil on absence, parse failure, version mismatch, or
    /// empty-tabs payload — core decides and logs the reason to stderr.
    static func load() -> Snapshot? {
        guard let cstr = copad_ffi_session_load() else { return nil }
        defer { copad_ffi_free_string(cstr) }
        let json = String(cString: cstr)
        guard let data = json.data(using: .utf8) else { return nil }
        do {
            return try JSONDecoder().decode(Snapshot.self, from: data)
        } catch {
            // Should not happen — core just produced this JSON. Surfacing
            // here means Swift's Codable and serde's `Session` drifted.
            FileHandle.standardError.write(
                Data("[copad] session decode (from core JSON) failed: \(error)\n".utf8),
            )
            return nil
        }
    }

    /// Returns whether the snapshot actually reached disk.
    ///
    /// The caller needs to know: history-blob GC runs against the COMMITTED
    /// session, and running it after a failed save would delete generations the
    /// previously committed session still references — destroying recoverable
    /// history rather than reclaiming dead files.
    @discardableResult
    static func save(_ snap: Snapshot) -> Bool {
        let encoder = JSONEncoder()
        let data: Data
        do {
            data = try encoder.encode(snap)
        } catch {
            FileHandle.standardError.write(
                Data("[copad] session encode failed: \(error)\n".utf8),
            )
            return false
        }
        let json = String(decoding: data, as: UTF8.self)
        let rc = json.withCString { copad_ffi_session_save($0) }
        if rc != 0 {
            let msg = copad_ffi_last_error().map { String(cString: $0) } ?? "<unknown>"
            FileHandle.standardError.write(
                Data("[copad] session save failed: \(msg)\n".utf8),
            )
            return false
        }
        return true
    }

    static func clear() {
        _ = copad_ffi_session_clear()
    }
}

// MARK: - Wire model

//
// PaneManager / TabViewController construct + consume these types in-
// process. The FFI surface only ever crosses serialized JSON, so the
// Swift types stay as the canonical in-memory shape for the rest of the
// macOS codebase.

extension Session {
    struct Snapshot: Codable, Equatable {
        let version: Int
        let tabs: [TabSnap]
        let currentTab: Int

        private enum CodingKeys: String, CodingKey {
            case version
            case tabs
            case currentTab = "current_tab"
        }
    }

    struct TabSnap: Codable, Equatable {
        let customTitle: String?
        let root: SplitSnap

        private enum CodingKeys: String, CodingKey {
            case customTitle = "custom_title"
            case root
        }
    }

    /// Wire-compatible with serde's `#[serde(tag = "type", rename_all =
    /// "snake_case")]` SplitSnap enum. A leaf carries typed `PaneContent`; a
    /// branch carries a normalized `ratio` (0..1). Manual Codable because
    /// Swift's auto-derived enum Codable picks a different on-disk layout.
    indirect enum SplitSnap: Equatable {
        case leaf(content: PaneContent)
        case branch(orientation: SplitOrientation, ratio: Float, first: SplitSnap, second: SplitSnap)
    }

    /// Wire-compatible with serde's `#[serde(tag = "kind", rename_all =
    /// "snake_case")]` PaneContent enum. Terminals persist only their cwd.
    enum PaneContent: Equatable {
        case terminal(cwd: String?)
        /// `pane` is the full browser state (tab list, active index, profile).
        /// Additive and omitted when nil, so a file this binary writes stays
        /// readable by an older one — which degrades to the single URL rather
        /// than failing (decision #100 / plan r2-I4).
        case webview(url: String, pane: BrowserPaneSnap?)
        case plugin(name: String, panelName: String?, version: String?)
        case cockpit
    }
}

extension Session.SplitSnap: Codable {
    private enum CodingKeys: String, CodingKey {
        case type, content, orientation, ratio, first, second
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try c.decode(String.self, forKey: .type)
        switch kind {
        case "leaf":
            self = .leaf(content: try c.decode(Session.PaneContent.self, forKey: .content))
        case "branch":
            let o = try c.decode(Session.SplitOrientation.self, forKey: .orientation)
            let r = try c.decode(Float.self, forKey: .ratio)
            let f = try c.decode(Session.SplitSnap.self, forKey: .first)
            let s = try c.decode(Session.SplitSnap.self, forKey: .second)
            self = .branch(orientation: o, ratio: r, first: f, second: s)
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .type, in: c,
                debugDescription: "unknown SplitSnap type: \(kind)",
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .leaf(content):
            try c.encode("leaf", forKey: .type)
            try c.encode(content, forKey: .content)
        case let .branch(orientation, ratio, first, second):
            try c.encode("branch", forKey: .type)
            try c.encode(orientation, forKey: .orientation)
            try c.encode(ratio, forKey: .ratio)
            try c.encode(first, forKey: .first)
            try c.encode(second, forKey: .second)
        }
    }
}

extension Session.PaneContent: Codable {
    private enum CodingKeys: String, CodingKey {
        case kind, cwd, url, name, version, pane
        case panelName = "panel_name"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(String.self, forKey: .kind) {
        case "terminal":
            self = .terminal(cwd: try c.decodeIfPresent(String.self, forKey: .cwd))
        case "webview":
            self = .webview(
                url: try c.decode(String.self, forKey: .url),
                // A malformed `pane` must not fail the whole session load — the
                // rest of the layout is still good, and losing one pane's tab
                // list beats starting from an empty window.
                pane: try? c.decodeIfPresent(BrowserPaneSnap.self, forKey: .pane),
            )
        case "plugin":
            self = .plugin(
                name: try c.decode(String.self, forKey: .name),
                panelName: try c.decodeIfPresent(String.self, forKey: .panelName),
                version: try c.decodeIfPresent(String.self, forKey: .version),
            )
        case "cockpit":
            self = .cockpit
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .kind, in: c, debugDescription: "unknown PaneContent kind",
            )
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .terminal(cwd):
            try c.encode("terminal", forKey: .kind)
            // Omit when nil to match serde's `skip_serializing_if = "Option::is_none"`.
            try c.encodeIfPresent(cwd, forKey: .cwd)
        case let .webview(url, pane):
            try c.encode("webview", forKey: .kind)
            try c.encode(url, forKey: .url)
            try c.encodeIfPresent(pane, forKey: .pane)
        case let .plugin(name, panelName, version):
            try c.encode("plugin", forKey: .kind)
            try c.encode(name, forKey: .name)
            try c.encodeIfPresent(panelName, forKey: .panelName)
            try c.encodeIfPresent(version, forKey: .version)
        case .cockpit:
            try c.encode("cockpit", forKey: .kind)
        }
    }
}

extension Session {
    /// Wire-facing orientation (the one in `SplitNode.swift` is
    /// renderer-facing). Both share the `"horizontal"`/`"vertical"`
    /// strings serde's `rename_all="lowercase"` produces.
    enum SplitOrientation: String, Codable, Equatable {
        case horizontal
        case vertical
    }
}
